[extractor/cinetecamilano] Add extractor

timendum opened the pull request:
### Description of your *pull request* and other information

</details>

Add new extractor or cinetecamilano.it
All movies are behind a login, but you can find one on [bugmenot](http://bugmenot.com/view/cinetecamilano.it).  
In the extractor, the auth is taken by cookie file.

Fixes #5031


<details open><summary>Template</summary> <!-- OPEN is intentional -->

<!--

# PLEASE FOLLOW THE GUIDE BELOW

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes `[ ]` relevant to your *pull request* (like [x])
- Use *Preview* tab to see how your *pull request* will actually look like

-->

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

pukkandan commented:
> ###
> * [x]  At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md?rgh-link-date=2022-10-18T12%3A14%3A14Z#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md?rgh-link-date=2022-10-18T12%3A14%3A14Z#yt-dlp-coding-conventions)

Pls do that again

timendum commented:
I've reviewed my changes, edited the formatting and changed release_timestamp, modified_timestamp and thumbnail fields.

If I missed something, please let me know.

I've not included test results, because I don't know how to perform testing with cookies.

pukkandan commented:
> I've not included test results, because I don't know how to perform testing with cookies.

https://github.com/yt-dlp/yt-dlp/blob/814bba3933ca36a79c68ac737b805cf25c407521/CONTRIBUTING.md?plain=1#L237-L243

Same applies for `cookies`

coletdjnz commented on yt_dlp/extractor/cinetecamilano.py:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    def _update_api_headers(self, headers):
+        token = self._get_cookies("https://www.cinetecamilano.it").get("cnt-token")
```
use single quotes
https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md?rgh-link-date=2022-10-18T12%3A14%3A14Z#quotes

coletdjnz commented on yt_dlp/extractor/cinetecamilano.py:
```diff
@@ -0,0 +1,56 @@
+                )
+            else:
+                raise e
+        if not film_json.get("success") or not film_json["success"] or not film_json.get("archive"):
```
If `success` does not exist this will fail, is that what you were wanting (not checking for `false`?)?

also this does the same thing
```suggestion
        if not film_json.get('success') or not film_json.get('archive'):
```

coletdjnz commented on yt_dlp/extractor/cinetecamilano.py:
```diff
@@ -0,0 +1,56 @@
+from json import JSONDecodeError
+from urllib.error import HTTPError
+from .common import InfoExtractor
+from ..utils import (
+    ExtractorError, float_or_none, parse_iso8601, strip_or_none, traverse_obj,
+    urljoin)
```
```suggestion
from ..utils import (
    ExtractorError, 
    float_or_none, 
    parse_iso8601, 
    strip_or_none, 
    traverse_obj,
    urljoin
)
```

pukkandan commented on yt_dlp/extractor/cinetecamilano.py:
```diff
@@ -0,0 +1,56 @@
+        if not film_json.get("success") or not film_json["success"] or not film_json.get("archive"):
+            raise ExtractorError("Video information not found")
+        archive = film_json["archive"]
+        info = {
+            "id": video_id,
+            "title": archive.get("title"),
+            "description": strip_or_none(archive.get("description")),
+            "duration": float_or_none(archive.get("duration"), invscale=60),
+            "release_timestamp": parse_iso8601(archive.get("updated_at"), delimiter=" "),
+            "modified_timestamp": parse_iso8601(archive.get("created_at"), delimiter=" "),
+        }
+        if traverse_obj(archive, "thumb", "src"):
+            info["thumbnail"] = urljoin(
+                url, archive["thumb"]["src"].replace("/public/", "/storage/"))
+        if traverse_obj(archive, "drm", "hls"):
+            info["formats"] = self._extract_m3u8_formats(
+                urljoin(url, archive["drm"]["hls"]), video_id)
+        return info
```
actually, move `formats` out of the return and make sure `self._sort_formats` is called

timendum commented:
Hi, thanks for the suggestion, I've accepted all of them, fixed one point and add a test case. I've also squashed all in one commit, for better order.

pukkandan commented:
> I've also squashed all in one commit, for better order.

That is unnecessary. In future, pls avoid force-pushing

pukkandan commented on yt_dlp/extractor/cinetecamilano.py:
```diff
@@ -0,0 +1,63 @@
+from json import JSONDecodeError
+from urllib.error import HTTPError
```
```suggestion
import json
import urllib.error
```
make necessary changes below

timendum replied:
I'm not sure about this one, should I import `json` and then catch `json.JSONDecodeError` ?

pukkandan replied:
yes

pukkandan commented on yt_dlp/extractor/cinetecamilano.py:
```diff
@@ -0,0 +1,63 @@
+    traverse_obj,
+    try_call,
+    try_get,
+    urljoin)
```
```suggestion
    urljoin,
)
```

pukkandan commented on yt_dlp/extractor/cinetecamilano.py:
```diff
@@ -0,0 +1,63 @@
+
+
+class CinetecaMilanoIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?cinetecamilano\.it/film/(?P<id>\d+)/?'
```
```suggestion
    _VALID_URL = r'https?://(?:www\.)?cinetecamilano\.it/film/(?P<id>\d+)'
```
it does nothing

pukkandan commented on yt_dlp/extractor/cinetecamilano.py:
```diff
@@ -0,0 +1,63 @@
+            'ext': 'mp4',
+            'title': "Il draghetto Gris\u00f9 (4 episodi)",
+            'release_date': '20220129',
+            'thumbnail': 're:.+\\.png',
```
```suggestion
            'thumbnail': r're:.+\.png',
```

pukkandan commented on yt_dlp/extractor/cinetecamilano.py:
```diff
@@ -0,0 +1,63 @@
+            'release_date': '20220129',
+            'thumbnail': 're:.+\\.png',
+            'description': 'md5:5328cbe080b93224712b6f17fcaf2c01',
+            'modified_date': "20200520",
```
```suggestion
            'modified_date': '20200520',
```

pukkandan commented on yt_dlp/extractor/cinetecamilano.py:
```diff
@@ -0,0 +1,63 @@
+        'info_dict': {
+            'id': '1942',
+            'ext': 'mp4',
+            'title': "Il draghetto Gris\u00f9 (4 episodi)",
```
```suggestion
            'title': 'Il draghetto Gris\u00f9 (4 episodi)',
```

pukkandan commented on yt_dlp/extractor/cinetecamilano.py:
```diff
@@ -0,0 +1,63 @@
+                    'Authorization': try_get(self._get_cookies('https://www.cinetecamilano.it'), lambda x: f"Bearer {x['cnt-token'].value}") or ''
+                })
+        except ExtractorError as e:
+            if (isinstance(e.cause, HTTPError) and e.cause.code == 500) or \
+                    isinstance(e.cause, JSONDecodeError):
```
```suggestion
            if ((isinstance(e.cause, HTTPError) and e.cause.code == 500)
                    or isinstance(e.cause, JSONDecodeError)):
```

pukkandan commented on yt_dlp/extractor/cinetecamilano.py:
```diff
@@ -0,0 +1,63 @@
+        except ExtractorError as e:
+            if (isinstance(e.cause, HTTPError) and e.cause.code == 500) or \
+                    isinstance(e.cause, JSONDecodeError):
+                raise ExtractorError(
+                    'This video is only available for registered users. You may want to use --cookies.',
+                    expected=True, cause=e.cause)
```
self.raise_login_required

pukkandan commented on yt_dlp/extractor/cinetecamilano.py:
```diff
@@ -0,0 +1,63 @@
+                f'https://www.cinetecamilano.it/api/catalogo/{video_id}',
+                video_id, headers={
+                    'Referer': url,
+                    'Authorization': try_get(self._get_cookies('https://www.cinetecamilano.it'), lambda x: f"Bearer {x['cnt-token'].value}") or ''
```
```suggestion
                    'Authorization': try_get(self._get_cookies('https://www.cinetecamilano.it'), lambda x: f'Bearer {x["cnt-token"].value}') or ''
```

pukkandan commented on yt_dlp/extractor/cinetecamilano.py:
```diff
@@ -0,0 +1,63 @@
+            'duration': float_or_none(archive.get('duration'), invscale=60),
+            'release_timestamp': parse_iso8601(archive.get('updated_at'), delimiter=' '),
+            'modified_timestamp': parse_iso8601(archive.get('created_at'), delimiter=' '),
+            'thumbnail': urljoin(url, try_call(lambda: archive['thumb']['src'].replace('/public/', '/storage/'))),
```
```suggestion
            'thumbnail': urljoin(url, try_get(archive, lambda x: x['thumb']['src'].replace('/public/', '/storage/'))),
```
no need to import both

pukkandan commented:
You reverted changes I made? 😕

timendum commented:
I've remade your suggestion by hand, I'm not so used to github workflows.

If you use a formatter (like autopep8 or yapf) can you share your settings?

Also, to load cookies for testing, the correct parameter in `local_parameters.json` is `cookiefile` (this is for future reference in case I will need it again)

pukkandan commented on yt_dlp/extractor/cinetecamilano.py:
```diff
@@ -0,0 +1,61 @@
+        video_id = self._match_id(url)
+        try:
+            film_json = self._download_json(
+                f'https://www.cinetecamilano.it/api/catalogo/{video_id}/?',
```
Why `/?`

pukkandan commented:
I do use autopep8, but 1335e5b26e6a327959fe6d58a0d467a293a58521 is manual changes. Don't worry about it. They really are nitpicks and I don't mind committing them myself. The only issue was when u (accidentally) reverted my commit
